ArmPkg/Drivers/CpuDxe/AArch64/Mmu.c

```diff
@@ -48,7 +48,8 @@ GetRootTranslationTableInfo (
   OUT UINTN  *RootTableEntryCount
   )
 {
-  *RootTableLevel      = (INTN)(T0SZ - MIN_T0SZ) / BITS_PER_LEVEL;
+  *RootTableLevel = ((T0SZ < MIN_T0SZ) && ArmLpa2Enabled ()) ?
+                    -1 : (INTN)(T0SZ - MIN_T0SZ) / BITS_PER_LEVEL;

```

Comment:
This should rely only on `T0SZ` - the fact that LPA2 needs to be enabled in order for level `-1` to be permitted is a separate matter, and returning `0` instead does not really help.

So better do something like

```
*RootTableLevel = (T0SZ < MIN_T0SZ) ? -1 : (INTN)(T0SZ - MIN_T0SZ) / BITS_PER_LEVEL;
ASSERT (*RootTableLevel >= 0 || ArmLpa2Enabled ());
```

Reply:
> Hi Ard, There is on corner case. When LPA2 feature enable, but DS but is not set (!ArmLpa2Enabled()), the page table level started from 0.
> 
> Your ASSERT() is rational but, the change of root level seems break above case.

That case will be broken anyway. Setting `TCR.T0SZ` to a value < 16 when `DS` unset is going to crash the system anyway, so whether the rootlevel is reported as `0` or `1` makes no difference whatsoever.